metagenomics.py

```diff
@@ -1143,6 +1144,162 @@ class KrakenBuildError(Exception):
+
+        
+        for k,taxa in sample_summary.items():
+            sample_summary[k] = collections.OrderedDict(sorted(taxa.items(), key=lambda item: (item[1][1]) , reverse=True)[:top_n_entries])
+
+            if len(list(sample_summary[k].items()))>0:
+                log.info("{f}: most abundant among {heading} at the {level} level: \"{name}\" with {reads} reads ({percent:.2%} of total); included since >{threshold} read{plural}".format(
+                                                                                                  f=f,
+                                                                                                  heading=k, 
+                                                                                                  level=taxlevel_focus, 
+                                                                                                  name=list(sample_summary[k].items())[0][0],
+                                                                                                  reads=list(sample_summary[k].items())[0][1][1],
+                                                                                                  percent=list(sample_summary[k].items())[0][1][0]/100.0,
+                                                                                                  threshold=count_threshold,
+                                                                                                  plural="s" if count_threshold>1 else "" )
+                )
+        samples[sample_name] = sample_summary
+    json_summary = json.dumps(samples, sort_keys=True, indent=4, separators=(',', ': '))
+    json_out.write(json_summary)
+    json_out.close()
+
+
+    if csv_out != None:
+
+        # if we're writing out at the same level as the query header
+        # write out the fractions and counts
+        if same_level or (top_n_entries==1 and no_hist):
+
+            fieldnames = set()
+            for sample, taxa in samples.items():
+                for heading,taxon in taxa.items():
+                    if len(taxon):
+                        for k in taxon.keys():
+                            fieldnames |= set([k+"-pt",k+"-ct"])
+
+            writer = csv.DictWriter(csv_out, fieldnames=["sample"]+sorted(list(fieldnames)))
+            writer.writeheader()
+
+            for sample, taxa in samples.items():
+                sample_dict = {}
+                sample_dict["sample"] = sample
+                for heading,taxon in taxa.items(): 
+                    for entry in taxon.keys():
+                        sample_dict[entry+"-pt"] = taxon[entry][0]
+                        sample_dict[entry+"-ct"] = taxon[entry][1]
+                writer.writerow(sample_dict)
+                
+
+            csv_out.close()
+
+        # otherwise write out a histogram
+        else:
+            count = 0
+            summary_counts = collections.defaultdict(dict)
+            for sample, totals in samples.items():
```

Comment:
Using pandas DataFrame might make these axis sum operations easier (and perhaps write out) - and also condense with above?

Reply:
Using pandas would help, but we don't currently have it as a dependency (and it seems a little silly to include just for this).

Reply:
"seems a little silly to include just for this" -- problem is, I think this every time I miss pandas, and there've been several over time.

Reply:
Many years ago I did a similar hesitant but eventual embrace of numpy/scipy for viral-ngs.. back before our conda-based installs, packages like this incurred a huge install-time cost on Travis, because `pip install` did a ton of compiling (about 15 mins worth of gcc/make). So I rewrote a bunch of stuff to avoid using it. Conda greatly reduces that install time now, and eventually numpy (but not scipy) was included as a conda dependency by something else (not sure exactly what), so I re-embraced numpy in the code. But I do think we need to carefully consider what the additional payload is for pandas+dependencies before dropping it in. We're already pretty heavyweight...